src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,218 @@
-
-        cartRepository.save(cart);
 
-        return CartMapper.cartEntityToDto(cart);
+        double total = cart.getTotal() - (initialQuantity * productDTO.getPrice()) + (quantity * productDTO.getPrice());
 
+        cart.setTotal(total);
 
+        return cartEntityToDto(cartRepository.save(cart));
     }
 
-    public CartDTO removeProductFromCart(String email, String productName, int quantity)
-            throws InvalidInputDataException, EntityNotFoundException {
+    public CartDTO removeProductFromCart(String email, String productName)
+            throws  EntityNotFoundException {
+        log.info("removeProductFromCart - process started");
 
-        Product product;
+        ProductDTO productDTO;
         try {
-            product = ProductMapper.DtoToProductEntity(productServiceClient.callGetProductByProductName(productName));
+            productDTO = productServiceClient.callGetProductByProductName(productName);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Product: " + productName + " not found in the Database!");
         }
 
-        User user;
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
-
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
-
-        if (productsInCart == 0) {
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
-        }
 
-        if (Math.toIntExact(productsInCart) < quantity) {
-            throw new InvalidInputDataException("You cannot remove more than " + productsInCart + " " + productName
-                    + " Products from the shopping cart!");
-        } else {
-            double productTotal = 0;
-
-            for (int i = 0; i < quantity; i++) {
-                productList.remove(product);
-                productTotal += product.getProductPrice();
-            }
+        Cart cart = cartRepository.findByUserId(user.getId()).orElseThrow(() ->
+         new EntityNotFoundException("Cart not found in the Database!"));
 
-            cart.setUser(user);
+        List<CartEntry> matchingEntries = cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).collect(Collectors.toList());
+        cart.getEntries().removeAll(matchingEntries);
+        cart.setTotal(cart.getTotal() - cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).count() * productDTO.getPrice());
```

Comment:
you should have done this before removing the entries from the cart. in this way 0 will be substracted from the total

Reply:
Done!